Draft: Archiving blog post (#4928)
kurrent-io/KurrentDB: docs/server/features/archiving_blog_post.md

> # Archiving: Hands on!

Reviewer: I think somewhere we should mention that archiving is simple to configure and use.  There's no specialness about archived data from the client point of view - it reads like any other data, just a bit slower due to the latency of the underlying object store.  This isn't just a cost play - it's also a scalability play and is really the foundation of separating storage from compute for KurrentDB.

Author: thanks! i've added these elements to the blog post

In this blog post, we'll walk you through using _Archiving_, our latest cool feature, available from KurrentDB 25.0.

> _Archiving_ can save you a tremendous amount of disk space on your KurrentDB nodes by moving old chunk files from the local disk to an external archive.

Reviewer 2: ```suggestion
_Archiving_ can save you a tremendous amount of disk space on your KurrentDB nodes by moving old chunk files from the local disk to an external archive such as S3.
```

Out of all the data accumulated over several years, recent data usually holds much more value than older data.
For example, in an e-commerce application, users are much more likely to view recent orders instead of orders they made ten years ago.

The recent data that is frequently accessed is called the _hot_ data, while the old data that is rarely accessed is called the _cold_ data.
When we keep the _hot_ and _cold_ data together on a fast, expensive disk, we're not using our resources optimally, resulting in higher costs.

What if we could separate the _cold_ data from the _hot_ data by making the _cold_ data still accessible but at slower speeds on a cheaper disk? This is the primary motivation behind the _Archiving_ feature!

## Defining what's _Hot_ and what's _Cold_
In KurrentDB, the tail of the transaction log, i.e., the latest chunk files, often naturally represent most of the _hot_ data, while old chunk files often represent most of the _cold_ data.
Thus, the basic idea is to keep the latest X chunks on the local disk while older chunks are stored in the archive based on a user-defined retention policy.

> The user-defined retention policy is simple: Keep X _days_ or Y _bytes_ of chunk data locally, whichever is larger.

Reviewer 2: ```suggestion
The user-defined retention policy is simple: Keep at least X _days_ and at least Y _bytes_ of chunk data locally.
```

To determine proper values for X & Y, there are some built-in metrics in KurrentDB.

You may also observe that there could be some _hot_ data present in the old chunks.

> This is handled through caching mechanisms to avoid hitting the archive frequently (work is in progress to improve this area).

Reviewer 2: There isn't a caching mechanism currently though

Author: updated, thanks

## The _Archiver_ node

> Using _Archiving_ requires adding a new node, called the _Archiver_ node, to the cluster. This node is solely responsible for uploading chunks to the archive.

Reviewer 2: ```suggestion
Using _Archiving_ requires adding a new node, called the _Archiver_ node, to the cluster. This node is responsible for uploading chunks to the archive.
```

## Demo time!
We'll use Linux for this demo, but the steps should be quite similar on Windows.

Enabling archiving is simple. Just add a few lines to your KurrentDB configuration file (located at `/etc/kurrentdb/kurrentdb.conf`) on all your nodes, including the _Archiver_ node:
```
Licensing:
  LicenseKey: <your key>

Archive:
  Enabled: true

  StorageType: S3
  S3:
    Region: eu-west-1
    Bucket: kurrentdb-cluster-123-archive

  RetainAtLeast:
    Days: 30
    LogicalBytes: 500000000
```

The configuration settings are straightforward - we first specify a license key and then enable archiving.
We then specify the storage type. Currently, we support only Amazon S3.  If you want us to support additional cloud providers, we would be happy to [hear from you](https://esdb.ideas.eventstore.com/)!
Then, we specify the S3 region and the S3 bucket where chunks will be archived.

Finally, we specify the retention policy.
Here, it means:
```
Keep the last chunks containing data that is not older than 30 days
or keep the latest 500MB of data, whichever is larger
```
We keep roughly two chunks of data or more depending on the data's age.

> ## S3 Authentication

Reviewer 2: It would be good to mention that it'll work if you just log in to the aws cli, no need to adjust the files

Author: updated, thanks

If you are using the default AWS profile, `default,` you don't need to complete these steps.

Specifying the `AWS_PROFILE` environment variable can be done as follows:
```
$ sudo systemctl edit kurrentdb
```

This will open up a text editor, where you must add these lines in the top section:
```
[Service]
Environment="AWS_PROFILE=kurrentdb-archive"
```

Save the file & exit.

Finally, reload the _systemd_ configuration by running:
```
$ sudo systemctl daemon-reload
```

We're now ready to spin up the nodes, including the _Archiver_ node!
```
$ sudo systemctl restart kurrentdb
```

## The _Archiver_ at work

After starting the nodes, nothing special happens at first sight on the cluster nodes.
If the archive settings were misconfigured, you will likely see errors as soon as the nodes start up.

If you look at the logs of the _Archiver_ node, you'll see that it has joined the cluster and started replicating chunk files from the leader as a normal node would.
You'll notice this line saying that the archive has a checkpoint of 0 (`0x0` in hex notation):
```
[INF] ArchiverService Archive is at checkpoint 0x0
```

> The _archive checkpoint_ represents the point in the KurrentDB transaction log before which everything has already been archived.

Reviewer 2: ```suggestion
The _archive checkpoint_ represents the point in the KurrentDB transaction log before which everything has already been added to the archive.
```

Therefore, chunks _before_ the _archive checkpoint_ can safely be deleted from all nodes.

> The checkpoint is measured in _bytes_. For example, a checkpoint of `0x10000000` means `268435456` bytes in decimal, which is exactly 256MiB (1 chunk)

Reviewer 2: ```suggestion
The checkpoint is measured in _bytes_. For example, a checkpoint of `0x10000000` means `268435456` bytes in decimal, which is exactly 256 MiB (1 chunk)
```

2025-03-20 11:30:46  268435456 chunk-000002.000001
```

One strange thing you'll notice above is that the chunk file names in the archive always have a suffix of `.000001`.
For example, `chunk-000000.000000` is stored as `chunk-000000.000001`, which looks similar to a _scavenged_ chunk.

The other strange thing you'll see is that if you have a _scavenged_ and _merged_ chunk that contains multiple logical chunks, it'll be stored as several chunk files in the archive.
For example, _scavenged_ chunk `chunk-000012.000003`, which contains chunks 12, 13, and 14, will be _unmerged_ and stored in the archive as `chunk-000012.000001`, `chunk-000013.000001`, and `chunk-000014.000001`.

This mechanism allows data to be read from the archive quickly. A constant suffix (`.000001`) and one chunk file per logical chunk allow any node to quickly locate the chunk file containing a specific _log position_ in the archive.
This has some (minor) implications: When you use _Archiving_, you cannot easily tell from a directory listing of the archive:

- whether a chunk was scavenged or not
- the number of times a chunk was scavenged (this information is local to a node anyway)

> After some time, all the chunks will eventually be uploaded to the archive.

Reviewer 2: ```suggestion
After some time, all the completed chunks will have been uploaded to the archive.
```

because of the retention policy for days
[DBG] SCAVENGING: ChunkRemover is keeping chunk "#1-1 (chunk-000001.000000)"
because of the retention policy for days
```

#### 30 days later...

We run a scavenge again 30 days later and sure enough, some of the chunks (0, 1, 2 & 3) are deleted from the node!
```
$ ls /var/lib/kurrentdb/
chaser.chk           chunk-000006.000001    index         writer.chk
chunk-000004.000000  chunk-000007.000000    proposal.chk
chunk-000005.000001  epoch.chk              truncate.chk
```

Notice that we had configured the archiving settings to keep only 500MB of data which should be roughly equivalent to 2 chunks but there are actually 4 chunks left.
That's because running a scavenge (two times) had closed the chunks 5 & 6 with a _scavenge point_ and they were thus not full of data.

You can now try to read some streams from a client or from the web UI, reads that go to the archive are handled transparently!

## Conclusion

This was a small introduction to _Archiving_.
Many questions may have popped up in your head, for example:

- How do I determine proper values for the retention policy?
- How do I scavenge the archive?
- How do backups work?

> 

Reviewer: Should we have a part 2 blog post that goes into these details?  I understand that they are in the documentation, but this may be useful content to also surface in another blog post as well.
